api/host_query_xjoin.py

```diff
@@ -185,14 +185,55 @@ def _query_filters(fqdn, display_name, hostname_or_id, insights_id, tags, stalen
 
     if filter:
         if filter.get("system_profile"):
-            if filter["system_profile"].get("sap_system"):
-                query_filters += build_sap_system_filters(filter["system_profile"]["sap_system"])
-            if filter["system_profile"].get("sap_sids"):
-                query_filters += build_sap_sids_filter(filter["system_profile"]["sap_sids"])
+            system_profile_filter = _build_system_profile_filter(filter["system_profile"])
+            if system_profile_filter != ():
+                query_filters += system_profile_filter
+        if filter.get("per_reporter_staleness"):
+            per_reporter_staleness_filter = _build_per_reporter_staleness_filter(filter["per_reporter_staleness"])
+            if per_reporter_staleness_filter != ():
+                query_filters += per_reporter_staleness_filter
 
     logger.debug(query_filters)
     return query_filters
 
 
+def _build_system_profile_filter(system_profile):
+    system_profile_filter = tuple()
+
+    if system_profile.get("sap_system"):
+        system_profile_filter += build_sap_system_filters(system_profile["sap_system"])
+    if system_profile.get("sap_sids"):
+        system_profile_filter += build_sap_sids_filter(system_profile["sap_sids"])
+
+    return system_profile_filter
+
+
+def _build_per_reporter_staleness_filter(per_reporter_staleness):
+    prs_dict_array = []
+
+    for reporter in per_reporter_staleness:
```

Comment:
```suggestion
    for reporter, props in per_reporter_staleness.items():
```
should drop duplicate sub-expressions nicely

Reply:
I agree... then you can just use `props["stale_timestamp"]` instead of `per_reporter_staleness[reporter]["stale_timestamp"]`, etc etc

Reply:
Done :+1: